File: backend/src/test/java/com/backend/config/TestRedisConfig.java
package com.backend.config;


import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;
import redis.embedded.RedisServer;

import java.io.IOException;
import java.net.ServerSocket;

@Configuration
@Profile("test")
public class TestRedisConfig {

    private RedisServer redisServer;
    private int redisPort;

    @PostConstruct
    public void startRedis() throws IOException {
        redisPort = findAvailableTcpPort();
        redisServer = new RedisServer(redisPort);
        redisServer.start();

        // Spring Boot Redis 설정에 반영
        System.setProperty("spring.data.redis.host", "localhost");
        System.setProperty("spring.data.redis.port", String.valueOf(redisPort));

        System.out.printf("Embedded Redis 서버 시작 (port: %d)%n", redisPort);
    }

    @PreDestroy
    public void stopRedis() throws IOException {
        if (redisServer != null && redisServer.isActive()) {
            redisServer.stop();
            System.out.println("Embedded Redis 서버 종료");
        }
    }

    private int findAvailableTcpPort() throws IOException {
        try (ServerSocket socket = new ServerSocket(0)) {
            socket.setReuseAddress(true);
            return socket.getLocalPort();
        }
    }
}

File: backend/src/test/java/com/backend/config/TestRedisConfig.kt
package com.backend.config

import jakarta.annotation.PostConstruct
import jakarta.annotation.PreDestroy
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import redis.embedded.RedisServer
import java.net.ServerSocket


@Configuration
@Profile("test")
class TestRedisConfig {
    private var redisServer: RedisServer? = null
    private var redisPort:Int = 0

    @PostConstruct
    fun startRedis() {
        redisPort = findAvailableTcpPort()
        redisServer = RedisServer(redisPort).also { it.start() }

        // Spring Boot Redis 설정에 반영
        System.setProperty("spring.data.redis.host", "localhost")
        System.setProperty("spring.data.redis.port", redisPort.toString())

        println("Embedded Redis 서버 시작 (port: ${redisPort})")
    }

    @PreDestroy
    fun stopRedis() {
        redisServer?.takeIf { it.isActive }?.stop()
        println("Embedded Redis 서버 종료")
    }


    private fun findAvailableTcpPort(): Int =
        ServerSocket(0).use { socket ->
            socket.reuseAddress = true
            socket.localPort

    }
}